File: Chapter_1/src/Section_1_2/gift1.java
package Section_1_2;
/*
ID: samueln1
LANG: JAVA
TASK: gift1
*/
import java.io.*;
import java.util.*;
public class gift1 {
	static Scanner in;
	static PrintWriter out;
	static int n;
	static LinkedHashMap<String, Integer> list;
	static StringBuilder sb;
	
	public static void main(String[] args) {
		try{
			in = new Scanner(new File("gift1.in"));
			out = new PrintWriter(new File("gift1.out"));
					
			init();
			out.print(solve());
			in.close();
			out.close();
		}catch(Exception e){
			System.out.println(e);
		}
		
	}
	
	private static void init(){
		list = new LinkedHashMap<String, Integer>();
		n = in.nextInt();
		in.nextLine();
		for(int i=0; i<n; i++){
			String name = in.nextLine();
			list.put(name, 0);
		}
	}
	
	private static String solve(){
		sb = new StringBuilder();
		for(int i=0; i<n; i++){
			String Giver = in.next();
			int money = in.nextInt();
			int d = in.nextInt();
			int dm=0;
			int left = 0;
			if(d==0){
				continue;
			}
			else{
				dm = money/d;
				left = money%d;
			}
			list.put(Giver,  list.get(Giver)-money+left);
			
			for(int j=0; j<d; j++){
				String r = in.next();
				list.put(r, list.get(r)+dm);
			}
			
		}
		for(String key: list.keySet()){
			sb.append(key).append(" ").append(list.get(key)).append("\n");
		}
		return sb.toString();
	}

}

File: Chapter_1/src/Section_1_2/gift1.py
"""
ID: samueln1
LANG: PYTHON3
TASK: gift1
"""

fin=open("gift1.in","r")
fout=open("gift1.out","w")
n=int(fin.readline()[:-1])

person = []
money = []
i=0
for i in range(n):
    person.append(fin.readline()[:-1])
    money.append(0)
i=0
for i in range(n):
    p = person.index(fin.readline()[:-1])
    l = fin.readline()[:-1].split()
    m=int(l[0])
    d = int(l[1])
    if(d==0):
        continue
    dm = m//d
    left = m%d
    
    for j in range(d):
        money[person.index(fin.readline()[:-1])]+=dm
    money[p]=money[p]-m+left

i=0
for i in range(n):
    print(person[i], str(money[i]))

fout.close()
fin.close()